Dprintf optimization

PeterSolMS opened the pull request:
Here's the initial implementation for putting the stress log into a memory mapped file and piping the dprintf calls to the stress log.

It's much faster than the dprintf implementation we normally and doesn't take a lock.

There is logic inside StressLog.cpp to switch to the memory mapped file when COMPLUS_StressLogFilename is set. Here's a set of environment variables I've used:

COMPlus_StressLogFilename=GCPerfSimStresslog.log
COMPlus_gcServer=1
COMPlus_LogLevel=9
COMPlus_StressLog=1
COMPlus_StressLogSize=10000000
COMPlus_TotalStressLogSize=1000000000

So this creates a max of 64 GB of stress log.

This mechanism also works for clrgc.dll, which accounts for some complications.

I have an analyzer to go with this that I'm still working on...

PeterSolMS committed:
delete backport.yml so I can push to master

PeterSolMS committed:
delete backport.yml so I can push to master

PeterSolMS committed:
Merge remote-tracking branch 'upstream/master'

PeterSolMS committed:
Undo unintended changes

PeterSolMS committed:
Undo unintended changes

PeterSolMS committed:
Merge remote-tracking branch 'upstream/master'

PeterSolMS committed:
Merge remote-tracking branch 'upstream/master'

PeterSolMS committed:
Merge remote-tracking branch 'upstream/master'

PeterSolMS committed:
Merge remote-tracking branch 'upstream/master'

PeterSolMS committed:
Merge remote-tracking branch 'upstream/master'

PeterSolMS committed:
Merge remote-tracking branch 'upstream/master'

PeterSolMS committed:
Merge remote-tracking branch 'upstream/master'

PeterSolMS committed:
Trying to make gc.cpp using stress log.

PeterSolMS committed:
Changes to support more arguments in LogMsg in a backward compatible way.

This also necessitates changes in SOS that I haven't done yet. However, stress logs containing no calls of more than 7 arguments would be dumped correctly by the current SOS - SOS would crash though with stress logs containing calls with more arguments.

PeterSolMS committed:
Merge remote-tracking branch 'upstream/master'

PeterSolMS committed:
Stresslog changes for memory mapped stress log.

PeterSolMS committed:
Merge remote-tracking branch 'upstream/master'

PeterSolMS committed:
Merge from upstream/master

PeterSolMS committed:
Merge from upstream/master

PeterSolMS committed:
Merge remote-tracking branch 'upstream/master'

PeterSolMS committed:
Merge branch 'master' of https://github.com/PeterSolMS/runtime-1

PeterSolMS committed:
Undo unintended Mono changes that somehow crept in.

PeterSolMS committed:
Adjust the memory-mapped stress log header so we have enough information for dumping from the memory-mapped file only.

PeterSolMS committed:
Merge remote-tracking branch 'upstream/master'

PeterSolMS committed:
Merge remote-tracking branch 'upstream/master'

PeterSolMS committed:
Merged with master, resolved conflicts (two moved files with changes).

PeterSolMS committed:
Stresslog changes to put stress log into a memory mapped file.

PeterSolMS committed:
Fixes for memory mapped stress log:
 - fixed region related oversight in dprintf
 - put thread id in the thread log chunks as well, to help with debugging
 - increased max file size to 32 GB.

PeterSolMS committed:
Merge remote-tracking branch 'upstream/master'

PeterSolMS committed:
Merge branch 'master' into dprintf-optimization

PeterSolMS committed:
Make dprintf via stress log mechanism work for clrgc.dll as well.

Added unrelated instrumentation to mark list equalization and sorting.

PeterSolMS committed:
Merge remote-tracking branch 'upstream/master'

PeterSolMS committed:
Merge remote-tracking branch 'upstream/master'

PeterSolMS committed:
Merge with master.

PeterSolMS committed:
Changes to make the mechanism more user-friendly:
- COMPLUS__StressLogFilename allows to set the name of the stress log file - if not set, stress log goes into in-memory buffers.
- Change processing of COMPLUS_TotalStressLogSize so that values >= 4 GB are supported

PeterSolMS committed:
Merge remote-tracking branch 'upstream/master'

PeterSolMS committed:
Merge branch 'master' into dprintf-optimization

Maoni0 commented on src/coreclr/gc/gc.cpp:
```diff
@@ -9283,8 +9283,14 @@ size_t gc_heap::sort_mark_list()
         return 0;
     }
 
+    uint64_t now = GetHighPrecisionTimeStamp();
```
since this is only used in dprintf you could put it under #ifdef TRACE_GC

Maoni0 reviewed (approved):
LGTM (the changes in gcpriv.h are still temporary right?)

PeterSolMS committed:
Implement putting dprintf log level into the stresslog.

PeterSolMS committed:
Stress log changes for GC:
- add dprintf level to stress log to allow for filtering based on it.
- make sure we only store new curPtr after a stress log message is completed

PeterSolMS committed:
Merge remote-tracking branch 'upstream/master'

PeterSolMS committed:
Merge branch 'master' into dprintf-optimization

PeterSolMS committed:
Merge with master, remove #ifdef around a dprintf that used to have too many args for stress log mechanism, but that we can handle now. Merge with master.

PeterSolMS committed:
Made #ifdefs consistent with master, removed #ifdef SIIMPLE_DPRINTF around a call that's now supported.

PeterSolMS committed:
Small fixes and changes.

PeterSolMS committed:
Merge remote-tracking branch 'upstream/master'

PeterSolMS committed:
Merge with Master

PeterSolMS committed:
More checking and robustness in memory-mapped stress log.

PeterSolMS committed:
Merge remote-tracking branch 'upstream/master'

PeterSolMS committed:
Merge remote-tracking branch 'upstream/master'

PeterSolMS committed:
Merge with Master

PeterSolMS committed:
Move stress log analyzer source into the src\coreclr\tools directory.

PeterSolMS committed:
Made StressLogAnalyzer use the stresslog.h version from inc - deleted the local copy.

PeterSolMS committed:
Moved dprintf format strings used by stress log analyzer to header files.

PeterSolMS committed:
Small fixes/improvements:
- teach stress log analyzer about background GC threads
- make message at the start of the mark phase consistent with other, similar ones - heap number goes first.

PeterSolMS committed:
Updated usage screen.

PeterSolMS committed:
Comment out definitions for USE_REGIONS and TRACE_GC.

PeterSolMS committed:
Merge branch 'dprintf-optimization' of https://github.com/PeterSolMS/runtime-1 into dprintf-optimization

PeterSolMS committed:
Consolidate header files.

PeterSolMS committed:
Adjust #ifdefs to make things compile for platforms/architectures other than Windows 64bit.

PeterSolMS committed:
Fix Linux build issue.

PeterSolMS committed:
Fix another Linux issue, get rid of mark list equalization timing dprintf.

PeterSolMS committed:
Fixed #endif on the wrong line on Linux.

Maoni0 commented on src/coreclr/gc/gcpriv.h:
```diff
@@ -344,7 +344,7 @@ class recursive_gc_sync;
 // but it works and is easier to debug for workstation GC
 // so turn it on for server GC, turn on for workstation GC if necessary
-#define FEATURE_CARD_MARKING_STEALING
+//#define FEATURE_CARD_MARKING_STEALING
```
temporary change?

PeterSolMS replied:
Right, did this for regions, thanks for catching it!

Maoni0 commented:
> #ifdef SIMPLE_DPRINTF

should we just get rid of SIMPLE_DPRINTF? the only reason I had it was just that stress log didn't take more than 7 params

---
Refers to: src/coreclr/gc/gcpriv.h:264 in ca2a702. [](commit_id = ca2a702d5b777ba7a62d36670cad7846f9416305, deletion_comment = False)

PeterSolMS commented:
> should we just get rid of SIMPLE_DPRINTF? the only reason I had it was just that stress log didn't take more than 7 params
Yes, eventually once we get comfy with the new scheme. Separate PR, I think.

PeterSolMS committed:
Undo temporary change to disable FEATURE_CARD_MARKING_STEALING.

PeterSolMS committed:
Merge branch 'dprintf-optimization' of https://github.com/PeterSolMS/runtime-1 into dprintf-optimization

PeterSolMS committed:
Added license headers to the StressLogAnalyzer source files, removed some German comments that VS put in.

AaronRobinsonMSFT commented:
Negatively impacted this support in https://github.com/dotnet/runtime/pull/50314. Working with @PeterSolMS offline on a way forward.
